src/Entities/Entity.cpp

```diff
@@ -1559,68 +1536,81 @@ bool cEntity::DetectPortal()
 
 
 
-bool cEntity::DoMoveToWorld(cWorld * a_World, bool a_ShouldSendRespawn, Vector3d a_NewPosition)
+void cEntity::DoMoveToWorld(const sWorldChangeInfo & a_WorldChangeInfo)
 {
-	UNUSED(a_ShouldSendRespawn);
-	ASSERT(a_World != nullptr);
+	ASSERT(a_WorldChangeInfo.m_NewWorld != nullptr);
 
-	if (GetWorld() == a_World)
+	if (m_World != a_WorldChangeInfo.m_NewWorld)
 	{
-		// Don't move to same world
-		return false;
+		// Tell others we are gone
+		m_World->BroadcastDestroyEntity(*this);
 	}
 
-	//  Ask the plugins if the entity is allowed to changing the world
-	if (cRoot::Get()->GetPluginManager()->CallHookEntityChangingWorld(*this, *a_World))
+	#ifdef _DEBUG
+		// Take note of old chunk coords
+		auto OldChunkCoords = cChunkDef::BlockToChunk(GetPosition());
+	#endif
+
+	// Set position to the new position
+	SetPosition(a_WorldChangeInfo.m_NewPosition);
+
+	if (a_WorldChangeInfo.m_SetPortalCooldown)
 	{
-		// A Plugin doesn't allow the entity to changing the world
-		return false;
+		m_PortalCooldownData.m_TicksDelayed = 0;
+		m_PortalCooldownData.m_ShouldPreventTeleportation = true;
+	}
+
+	if (GetWorld() == a_WorldChangeInfo.m_NewWorld)
+	{
+		// Moving to same world, don't need to remove from world
+		return;
 	}
 
 	// Stop ticking, in preperation for detaching from this world.
 	SetIsTicking(false);
 
-	// Tell others we are gone
-	GetWorld()->BroadcastDestroyEntity(*this);
+	a_WorldChangeInfo.m_NewWorld->BroadcastSpawnEntity(*this);
 
-	// Take note of old chunk coords
-	auto OldChunkCoords = cChunkDef::BlockToChunk(GetPosition());
+	// Add to new world and removal from the old one
+	LOGD("Warping entity #%i (%s) from world \"%s\" to \"%s\". Source chunk: (%d, %d) ",
+		this->GetUniqueID(), this->GetClass(),
+		m_World->GetName().c_str(), a_WorldChangeInfo.m_NewWorld->GetName().c_str(),
+		OldChunkCoords.m_ChunkX, OldChunkCoords.m_ChunkZ
+	);
+	a_WorldChangeInfo.m_NewWorld->AddEntity(m_World->RemoveEntity(*this));
+	cRoot::Get()->GetPluginManager()->CallHookEntityChangedWorld(*this, *m_World);
+}
 
-	// Set position to the new position
-	SetPosition(a_NewPosition);
 
-	// Stop all mobs from targeting this entity
-	// Stop this entity from targeting other mobs
-	if (this->IsMob())
-	{
-		cMonster * Monster = static_cast<cMonster*>(this);
-		Monster->SetTarget(nullptr);
-		Monster->StopEveryoneFromTargetingMe();
-	}
 
-	// Queue add to new world and removal from the old one
-	cWorld * OldWorld = GetWorld();
-	SetWorld(a_World);  // Chunks may be streamed before cWorld::AddPlayer() sets the world to the new value
```

Comment:
Does SetWorld need to still be called for cEntity? (or, is it and I just missed it)

(at any rate, old plugins might expect the entity's world to be the new world immediately after calling MoveToWorld)

Reply:
It is set in `cWorld::AddEntity`.  Setting immediately from `MoveToWorld` could be up for discussion though.

Reply:
This is ostensibly the PR that first introduced that particular call to SetWorld:
https://github.com/cuberite/cuberite/pull/1061

This PR later added the comment when it was moved into the queue system:
https://github.com/cuberite/cuberite/pull/2986

Reply:
Note that at the time of #1061 `cWorld::AddEntity` didn't actually call SetWorld immediately.

Reply:
It may not be an issue anymore, then.